engines/wintermute/base/gfx/opengl/base_render_opengl3d.cpp

```diff
@@ -0,0 +1,384 @@
+
+	*(vertices) = r;
+	*(vertices + 1) = g;
+	*(vertices + 2) = b;
+	*(vertices + 3) = a;
+	*(vertices + vertex_size) = r;
+	*(vertices + vertex_size + 1) = g;
+	*(vertices + vertex_size + 2) = b;
+	*(vertices + vertex_size + 3) = a;
+	*(vertices + 2 * vertex_size) = r;
+	*(vertices + 2 * vertex_size + 1) = g;
+	*(vertices + 2 * vertex_size + 2) = b;
+	*(vertices + 2 * vertex_size + 3) = a;
+	*(vertices + 3 * vertex_size) = r;
+	*(vertices + 3 * vertex_size + 1) = g;
+	*(vertices + 3 * vertex_size + 2) = b;
+	*(vertices + 3 * vertex_size + 3) = a;
+
+	glDisable(GL_DEPTH_TEST);
+	glEnable(GL_BLEND);
+	glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+	glBindTexture(GL_TEXTURE_2D, 0);
+
+	glEnableClientState(GL_VERTEX_ARRAY);
+	glEnableClientState(GL_COLOR_ARRAY);
+
+	glVertexPointer(3, GL_FLOAT, vertex_size, vertices + 4);
+	glColorPointer(4, GL_UNSIGNED_BYTE, vertex_size, vertices);
+
+	glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
+
+	setup2D(true);
+}
+
+bool Wintermute::BaseRenderOpenGL3D::drawLine(int x1, int y1, int x2, int y2, uint32 color) {
+	return true;
+}
+
+bool Wintermute::BaseRenderOpenGL3D::drawRect(int x1, int y1, int x2, int y2, uint32 color, int width) {
+	return true;
+}
+
+bool Wintermute::BaseRenderOpenGL3D::setProjection() {
+	// is the viewport already set here?
+	float viewportWidth = _viewportRect.right - _viewportRect.left;
+	float viewportHeight = _viewportRect.bottom - _viewportRect.top;
+
+	float horizontal_view_angle = M_PI * 0.5f;
+	float aspect_ratio = float(viewportHeight) / float(viewportWidth);
+	float near_plane = 1.0f;
+	float far_plane = 1000.0f;
+	float right = near_plane * tanf(horizontal_view_angle * 0.5f);
+
+	glMatrixMode(GL_PROJECTION);
+	glLoadIdentity();
+	glFrustum(-right, right, -right * aspect_ratio, right * aspect_ratio, near_plane, far_plane);
+	glMatrixMode(GL_MODELVIEW);
+
+	return true;
+}
+
+bool Wintermute::BaseRenderOpenGL3D::setProjection2D() {
+	glMatrixMode(GL_PROJECTION);
+	glLoadIdentity();
+	glOrtho(0, _viewportRect.width(), 0, _viewportRect.height(), -1.0, 100.0);
+	glMatrixMode(GL_MODELVIEW);
+	glLoadIdentity();
+	return true;
+}
+
+bool Wintermute::BaseRenderOpenGL3D::windowedBlt() {
+	return true;
+}
+
+bool Wintermute::BaseRenderOpenGL3D::fill(byte r, byte g, byte b, Common::Rect *rect) {
+	glClearColor(float(r) / 255.0f, float(g) / 255.0f, float(b) / 255.0f, 1.0f);
+	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
+	return true;
+}
+
+void Wintermute::BaseRenderOpenGL3D::onWindowChange() {
+}
+
+bool Wintermute::BaseRenderOpenGL3D::initRenderer(int width, int height, bool windowed) {
+	_windowed = windowed;
+	_width = width;
+	_height = height;
+
+	setViewport(0, 0, width, height);
+
+	_active = true;
+	// setup a proper state
+	setup2D(true);
+	return true;
+}
+
+bool Wintermute::BaseRenderOpenGL3D::flip() {
+	g_system->updateScreen();
+	return true;
+}
+
+bool Wintermute::BaseRenderOpenGL3D::indicatorFlip() {
+	return true;
+}
```

Comment:
Note: IndicatorFlip is an optimization for the rendering that happens during save-game saving, as the progress indicator that gets drawn during loading would slow down the loading depending on the frame rate (since it all happens in the same thread). Just in case you wondered what the deal is.